tour_model_eval/load_predict.py

```diff
@@ -44,13 +45,29 @@ def predict_labels(trip):
     trip_loc_feat = trip_feat[0:4]
     try:
         # load locations of bins(1st round of clustering)
+        # e.g.{'0': [[start lon1, start lat1, end lon1, end lat1],[start lon, start lat, end lon, end lat]]}
+        # another explanation: -'0': label from the 1st round
+        #                      - the value of key '0': all trips that in this bin
+        #                      - for every trip: the coordinates of start/end locations
         bin_locations = loadModelStage('locations_' + str(user))[0]
+
         # load models from the 2nd round of clustering
+        # we use Kmeans to build the model in the previous model building step
+        # assume that we have 2 clusters from the 1st round(that means 2 bins),
+        # the following is an example of the saved models.
+        # e.g. {'0': KMeans(n_clusters=2, random_state=0), '1': KMeans(n_clusters=5, random_state=0)}
         models = loadModelStage('models_' + str(user))[0]
+
         # load user labels in all clusters
+        # assume that we have 1 cluster(bin) from the 1st round of clustering, which has label '0',
+        # and we have 1 cluster from the 2nd round, which has label '1'
+        # the value of key '0' contains all 2nd round clusters
+        # the value of key '1' contains all user labels and probabilities in this cluster
+        # e.g. {'0': [{'1': [{'labels': {'mode_confirm': 'shared_ride', 'purpose_confirm': 'home', 'replaced_mode': 'drove_alone'}}]}]}
```

Comment:
is there a reason why this is not just

```
{'01': [{'labels': {'mode_confirm': 'shared_ride', 'purpose_confirm': 'home', 'replaced_mode': 'drove_alone'}}]}]}
```

Also, I want to confirm that you are in fact storing multiple possible labels and the associated probabilities, because that is not what I see here.

I would expect this to actually be

```
[{'labels': {'mode_confirm': 'shared_ride', 'purpose_confirm': 'home', 'replaced_mode': 'drove_alone'}, "p": 1.0}]
```

Reply:
I have two rounds of clustering. Let's say the final result has 2 layers. The first layer is result from the first round, the second layer is the result from the second round. So, in the first layer, every bin contains all second round labels and associate user labels and `p` in each second round cluster

What you expect is my final result. When the new trip matches a 2nd round cluster, then the associate labels and `p` would look like that.

Reply:
> I have two rounds of clustering. Let's say the final result has 2 layers. The first layer is result from the first round, the second layer is the result from the second round. 

I understand but you are not addressing my comment. Why is it not enough to have the label be the concatenation of the first and second rounds - e.g. `{'0': {'1': ...}` to `{'01': {....`

Reply:
I don't have the model that labeled '1', '2', '11', '12'... since the first round was not from any library. What Kmeans did was label the 2nd round cluster [0,1,2,3,...]. What I got using Kmeans `predict` is like [1]. Therefore, it is better to separate into 2 layers instead of getting concatenation of labels. 
I have to get the 1st label anyway, then `predict` gives me the 2nd label. I just use the 2nd label to get the result. Otherwise, in the labels saving step, it would be complicated. Also, in the predict step, I have to concatenate the labels again.

Reply:
None of the reasons make sense to me ("otherwise in the labels saving step, it would be complicated" Why?), but it is also not worth arguing further about this.

Although you are not able to justify it, your design decision is not a-priori incorrect, so I'm going to move on.